File: BEmu/BloombergTypes/Session.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="BloombergTypes/Session.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    using System.Threading;

    public class Session
    {
        private SessionOptions _sessionOptions;

        #pragma warning disable 0414 //disables the "is assigned but its value is never used" warning
        private enum SessionResponseType { sync, async }
        private readonly SessionResponseType _sessionResponse;

        private enum SessionUriType { undefined, refData, mktData }
        private SessionUriType _sessionUri = SessionUriType.undefined;

        private enum SessionStateType { initialized, started, serviceOpened }
        private SessionStateType _sessionState = SessionStateType.initialized;
        #pragma warning restore 0414

        private readonly Queue<Request> _sentRequests;
        private readonly EventHandler _asyncHandler;
        private CorrelationID _asyncOpenCorrelation;
        private readonly List<Subscription> _subscriptions;
        private readonly Timer _marketSimulatorTimer;
        private readonly object _syncroot = new object();

        #region SYNC
        public Session(SessionOptions sessionOptions) //for ReferenceData and HistoricalData (sync)
        {
            this._sessionResponse = SessionResponseType.sync;
            this._sessionOptions = sessionOptions;
            this._sentRequests = new Queue<Request>();
            this._asyncHandler = null;
            this._subscriptions = null;
            this._marketSimulatorTimer = null;
        }

        public bool Start()
        {
            this._sessionState = SessionStateType.started;
            return true;
        }

        public bool OpenService(string uri)
        {
            this._sessionState = SessionStateType.serviceOpened;
            return true;
        }

        public Service GetService(string uri)
        {
            if (uri == "//blp/refdata")
            {
                this._sessionUri = SessionUriType.refData;
                return new ServiceRefData();
            }
            else
                throw new NotImplementedException(string.Format("BEmu.Session.GetService: Service {0} not supported", uri));
        }

        public CorrelationID SendRequest(Request request, CorrelationID correlationId)
        {
            if (request is HistoricalDataRequest.RequestHistoric)
            {
                if (!((HistoricalDataRequest.RequestHistoric)request).DtStart.HasValue)
                    throw new ArgumentException("Historic requests must have start dates");
            }

            request.correlationId = correlationId;
            this._sentRequests.Enqueue(request);
            return correlationId;
        }

        public Event NextEvent()
        {
            bool isLastRequest = this._sentRequests.Count == 1;
            return Event.EventFactory(this._sentRequests.Dequeue(), isLastRequest);
        }
        #endregion

        #region ASYNC
        public Session(SessionOptions sessionOptions, EventHandler handler) //for MarketData (async)
        {
            this._sessionResponse = SessionResponseType.async;
            this._sessionOptions = sessionOptions;
            this._sentRequests = null;
            this._asyncHandler = handler;
            this._subscriptions = new List<Subscription>();
            this._marketSimulatorTimer = new Timer(new TimerCallback(this.MarketSimulatorTimerStep), null, (int)TimeSpan.FromSeconds(0).TotalMilliseconds, Timeout.Infinite);
        }

        public bool StartAsync()
        {
            this._sessionState = SessionStateType.started;
            MarketDataRequest.EventMarket evtSessionStatus = new MarketDataRequest.EventMarket(Event.EventType.SESSION_STATUS, null, null);
            MarketDataRequest.EventMarket evtServiceStatus = new MarketDataRequest.EventMarket(Event.EventType.SERVICE_STATUS, new CorrelationID(), null);

            if (this._asyncHandler != null)
            {
                this._asyncHandler(evtSessionStatus, this);
                this._asyncHandler(evtServiceStatus, this);
            }
            
            return true;
        }

        public void Stop()
        {
            this._marketSimulatorTimer.Change(Timeout.Infinite, Timeout.Infinite);
        }

        public void Dispose()
        {
            this._marketSimulatorTimer.Dispose();
        }

        public void OpenServiceAsync(string uri, CorrelationID correlationId)
        {
            this._sessionUri = SessionUriType.mktData;
            this._sessionState = SessionStateType.serviceOpened;
            this._asyncOpenCorrelation = correlationId;
        }

        public void Subscribe(IList<Subscription> subscriptionList)
        {
            lock (this._syncroot) //protect _subscriptions
                this._subscriptions.AddRange(subscriptionList);

            MarketDataRequest.EventMarket evtSubStatus = new MarketDataRequest.EventMarket(Event.EventType.SUBSCRIPTION_STATUS, null, subscriptionList);
            if (this._asyncHandler != null)
            {
                this._asyncHandler(evtSubStatus, this);
            }
        }

        public void Cancel(CorrelationID corr)
        {
            lock (this._syncroot) //protect _subscriptions
            {
                for (int i = this._subscriptions.Count - 1; i >= 0; i--)
                {
                    if (this._subscriptions[i].CorrelationID.Value == corr.Value && this._subscriptions[i].CorrelationID.IsInternal == corr.IsInternal)
                    {
                        MarketDataRequest.EventMarket evtSubCancel = new MarketDataRequest.EventMarket(Event.EventType.SUBSCRIPTION_STATUS, this._subscriptions[i]);
                        if (this._asyncHandler != null)
                        {
                            this._asyncHandler(evtSubCancel, this);
                        }
                        this._subscriptions.RemoveAt(i);
                    }
                }
            }
        }

        private void MarketSimulatorTimerStep(object arg)
        {
            int? conflationIntervalInMilleseconds = null;

            lock (this._syncroot) //protect _subscriptions
            {
                List<Subscription> subsToUse = new List<Subscription>();                

                foreach (var item in this._subscriptions)
                {
                    if (Types.RandomDataGenerator.ShouldIncludeQuote()) //70% chance that I'll send a new quote for the current subscription (after the first response which contains all tickers)
                        subsToUse.Add(item);

                    if (item.ConflationInterval.HasValue)
                        conflationIntervalInMilleseconds = item.ConflationInterval.Value * 1000;
                }

                MarketDataRequest.EventMarket evt = new MarketDataRequest.EventMarket(Event.EventType.SUBSCRIPTION_DATA, null, subsToUse);

                if (this._asyncHandler != null)
                    this._asyncHandler(evt, this);
            }

            this._marketSimulatorTimer.Change(conflationIntervalInMilleseconds.GetValueOrDefault((int)Types.RandomDataGenerator.TimeBetweenMarketDataEvents().TotalMilliseconds), Timeout.Infinite);
        }
        #endregion

    }
}

File: BEmuJava/bemu/src/main/java/com/bemu/BEmu/Session.java
//------------------------------------------------------------------------------
// <copyright project="BEmuJava" file="BEmu.Session.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bemu.BEmu;

import com.bemu.BEmu.SessionOptions;
import com.bemu.BEmu.CorrelationID;
import com.bemu.BEmu.Request;
import com.bemu.BEmu.Event;
import com.bemu.BEmu.types.RandomDataGenerator;
import com.bemu.BEmu.MarketDataRequest.EventMarket;

import java.util.ArrayList;
import java.util.List;
import java.util.Queue;
import java.util.LinkedList;
import java.util.concurrent.atomic.AtomicBoolean;

public class Session {
	@SuppressWarnings("unused")
	private SessionOptions _sessionOptions;
	
	@SuppressWarnings("unused")
	private SessionResponseType _sessionResponse;
	private enum SessionResponseType { sync, async };
	
	@SuppressWarnings("unused")
	private SessionUriType _sessionUri = SessionUriType.undefined;
	private enum SessionUriType { undefined, refData, mktData };
	
	@SuppressWarnings("unused")
	private SessionStateType _sessionState = SessionStateType.initialized;
	private enum SessionStateType { initialized, started, serviceOpened };
	
	private final Queue<Request> _sentRequests;
    
	////////////////////////////////////////
	// Start Sync
	////////////////////////////////////////
	
	public Session(SessionOptions sessionOptions)
	{
		this._sessionOptions = sessionOptions;
		this._sessionResponse = SessionResponseType.sync;
		this._sentRequests = new LinkedList<Request>();
        this._asyncHandler = null;
        //this._marketSimulatorTimer = null;
        this._marketSimulatorTypeClass = null;
        this._subscriptions = null;
	}
	
    public boolean start()
    {
        this._sessionState = SessionStateType.started;
        return true;
    }
    
    public boolean openService(String uri)
    {
        this._sessionState = SessionStateType.serviceOpened;
        return true;
    }
    
    public Service getService(String uri) throws Exception
    {
        if (uri.equals("//blp/refdata"))
        {
            this._sessionUri = SessionUriType.refData;
            return new ServiceRefData();
        }
        else
            throw new Exception(String.format("BEmu.Session.GetService: Service %s not supported", uri));
    }
    
    public CorrelationID sendRequest(Request request, CorrelationID correlationId)
    {
        request.correlationId(correlationId);
        this._sentRequests.add(request);
        return correlationId;
    }
    
    public Event nextEvent() throws Exception
    {
    	boolean isLastRequest = this._sentRequests.size() == 1;
    	Request next = this._sentRequests.poll();
    	return Event.EventFactory(next, isLastRequest);
    }
    
	////////////////////////////////////////
	// End Sync
	////////////////////////////////////////
    

    
	////////////////////////////////////////
	// Start Async
	////////////////////////////////////////

    private final EventHandler _asyncHandler;
    private final SubscriptionList _subscriptions;
    private final Object _syncroot = new Object();
    private final MarketSimulatorTypeClass _marketSimulatorTypeClass;
    private AtomicBoolean _isMarketSimulatorRunning;
    
    @SuppressWarnings("unused")
	private CorrelationID _asyncOpenCorrelation;
    
    public Session(SessionOptions sessionOptions, EventHandler handler) //for MarketData (async)
    {
        this._sessionResponse = SessionResponseType.async;
        this._sessionOptions = sessionOptions;
        this._sentRequests = null;
        this._asyncHandler = handler;
        this._subscriptions = new SubscriptionList(); 
        this._isMarketSimulatorRunning = new AtomicBoolean(false);
        this._marketSimulatorTypeClass = new MarketSimulatorTypeClass(this);
    }
    
    public void openServiceAsync(String url, CorrelationID correlationId)
    {
    	this._sessionUri = SessionUriType.mktData;
    	this._sessionState = SessionStateType.serviceOpened;
        this._asyncOpenCorrelation = correlationId;
    }
    
    public void subscribe(SubscriptionList subscriptionList) throws Exception
    {
    	synchronized (this._syncroot) //protect _subscriptions
    	{
    		List<Subscription> slist = subscriptionList.list();
    		for(int i = 0; i < slist.size(); i++)
    		{
    			this._subscriptions.add(slist.get(i));
    		}
    	}

        Event.EventType evttSub = new Event.EventType(Event.EventType.Constants.SUBSCRIPTION_STATUS);
        EventMarket evtSubStatus = new EventMarket(evttSub, null, subscriptionList.list());
        if (this._asyncHandler != null)
        {
            this._asyncHandler.processEvent(evtSubStatus, this);
        }
    }
    
    public boolean startAsync() throws Exception
    {
        this._sessionState = SessionStateType.started;
        this._isMarketSimulatorRunning.set(true);

        Event.EventType evttSession = new Event.EventType(Event.EventType.Constants.SESSION_STATUS);
        Event.EventType evttService = new Event.EventType(Event.EventType.Constants.SERVICE_STATUS);
        
        EventMarket evtSessionStatus = new EventMarket(evttSession, null, null);
        EventMarket evtServiceStatus = new EventMarket(evttService, new CorrelationID(), null);

        if (this._asyncHandler != null)
        {
            this._asyncHandler.processEvent(evtSessionStatus, this);
            this._asyncHandler.processEvent(evtServiceStatus, this);
        }
        
        this._marketSimulatorTypeClass.run();
        
        return true;
    }

    public void stop()
    {
        this._isMarketSimulatorRunning.set(false);
    }
    
    private class MarketSimulatorTypeClass extends java.lang.Thread
    {
    	private final Session _session;
    	
    	public MarketSimulatorTypeClass(Session session)
    	{
    		this._session = session;
    	}
    	
    	public void run()
    	{
    		boolean interrupted = false;
    		while(this._session._isMarketSimulatorRunning.get() && !interrupted)
    		{
                Long conflationIntervalInMilleseconds = null;

                List<Subscription> subsToUse = new ArrayList<Subscription>();  
                synchronized (this._session._syncroot) //protect _subscriptions
                {
                	List<Subscription> subscriptions = this._session._subscriptions.list();
                    
                	for(int i = 0; i < subscriptions.size(); i++)
                	{
                        if (RandomDataGenerator.shouldIncludeQuote()) //70% chance that I'll send a new quote for the current subscription (after the first response which contains all tickers)
                        	subsToUse.add(subscriptions.get(i));

                        if (subscriptions.get(i).conflationInterval() != null)
                            conflationIntervalInMilleseconds = (long)(subscriptions.get(i).conflationInterval() * 1000L);
                	}
                }
                
            	Event.EventType evtt = new Event.EventType(Event.EventType.Constants.SUBSCRIPTION_DATA);            	
                try
                {
    				EventMarket evt = new EventMarket(evtt, null, subsToUse);
    				if(this._session._asyncHandler != null)
    					this._session._asyncHandler.processEvent(evt, this._session);
    			}
                catch (Exception e)
                {
                	//do nothing
    			}
                
                if(conflationIntervalInMilleseconds == null)
                	conflationIntervalInMilleseconds = RandomDataGenerator.millisecondsBetweenMarketDataEvents();
                try
                {
                	Thread.sleep(conflationIntervalInMilleseconds);
				}
                catch (Exception e)
                {
                	interrupted = true;
				}
    		}
    	}
    }
    
	////////////////////////////////////////
	// End Async
	////////////////////////////////////////
    
}
